dremio_toolkit/create_snapshot.py

```
@@ -44,13 +34,26 @@ def create_snapshot(env_api: EnvApi, logger: Logger, output_file: str, datetime_
     arg_parser.add_argument("-u", "--user", help="User name. User must be a Dremio admin.", required=True)
     arg_parser.add_argument("-p", "--password", help="User password.", required=True)
     arg_parser.add_argument("-o", "--output-file", help="Json file name to save Dremio environment.", required=True)
+    arg_parser.add_argument("-r", "--report-file", help="CSV file name for the exception report.", required=False)
+    arg_parser.add_argument("-e", "--report-delimiter", help="Delimiter to use in the exception report. Default is tab.", required=False, default='\t')
     arg_parser.add_argument("-l", "--log-level", help="Set Log Level to DEBUG, INFO, WARN, ERROR.",
                             choices=['ERROR', 'WARN', 'INFO', 'DEBUG'], default='WARN')
     arg_parser.add_argument("-v", "--verbose", help="Set Log to verbose to print object definitions instead of object IDs.",
                             required=False, default=False, action='store_true')
+    arg_parser.add_argument("-f", "--log-file", help="Set Log to write to a specified file instead of STDOUT.",
+                            required=False)
     args = arg_parser.parse_args()
 
-    logger = Logger(level=args.log_level, verbose=args.verbose)
+    if args.report_file is None:
+        print("report-file argument has not been specified. Exception report will not be produced.")
+
+    logger = Logger(level=args.log_level, verbose=args.verbose, log_file=args.log_file)
+
     env_api = EnvApi(args.dremio_environment_url, args.user, args.password, logger)
+    env_reader = EnvReader(env_api, logger)
+    env_def = env_reader.read_dremio_environment(report_file=args.report_file, delimiter=args.report_delimiter)
+    EnvFileWriter.save_dremio_environment(env_def, args.output_file)
 
-    create_snapshot(env_api=env_api, logger=logger, output_file=args.output_file, datetime_utc=datetime.utcnow())
+    logger.finish_process_status_reporting()
+    if logger.get_error_count() > 0:
+        exit(1)
```

that's kind of a "silent fail" from user's perspective, maybe we should at least print information that there were some errors?

---

The error count gets increased by calling logger.error(). This function creates a log entry. Error count is also printed with the progress on the console. exit(1) is only to report error to the calling process.

---

ok, makes sense :)